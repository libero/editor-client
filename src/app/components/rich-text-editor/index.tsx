import React from 'react';
import { EditorView } from 'prosemirror-view';
import { EditorState, Transaction, TextSelection } from 'prosemirror-state';
import { debounce, get } from 'lodash';
import { DOMSerializer, Slice } from 'prosemirror-model';
import 'prosemirror-example-setup/style/style.css';
import 'prosemirror-menu/style/menu.css';
import { SectionContainer, SectionContainerVariant } from 'app/components/section-container';
import { ReferenceCitationNodeView } from 'app/components/reference-citation-editor-popup';
import { ComponentWithId } from 'app/utils/types';
import { BoxTextNodeView } from 'app/components/box-text';
import { hasParentNodeOf } from 'app/utils/view.utils';
import { LinkNodeView } from 'app/components/link-editor-popup';
import { FigureNodeView } from 'app/components/figure';
import { FigureCitationNodeView } from 'app/components/figure-citation';

/*
 * available classes:
 *   - root: applied to SectionContainer
 * */

export interface RichTextEditorProps {
  editorState: EditorState;
  label?: string;
  classes?: Record<string, string>;
  name?: string;
  isActive: boolean;
  variant?: SectionContainerVariant;
  onChange?: (change: Transaction, name: string) => void;
  onFocus?: (state: EditorState, name: string) => void;
  onBlur?: (state: EditorState, name: string) => void;
}

const restoreSelection = debounce((editorView, from, to) => {
  if (editorView.state.selection) {
    const $from = editorView.state.doc.resolve(from);
    const $to = editorView.state.doc.resolve(to);
    const change = editorView.state.tr.setSelection(new TextSelection($from, $to));
    editorView.dispatch(change);
  }
}, 50);

export class RichTextEditor extends React.Component<ComponentWithId<RichTextEditorProps>> {
  private options;
  public editorView: EditorView;

  constructor(props: ComponentWithId<RichTextEditorProps>) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.options = {
      nodeViews: {
        refCitation(node, view, getPos) {
          return new ReferenceCitationNodeView(node, view, getPos);
        },
        boxText: (node, view, getPos) => {
          return new BoxTextNodeView(node, view, getPos);
        },
        link(node, view) {
          return new LinkNodeView(node, view);
        },
        figure: (node, view, getPos) => {
          return new FigureNodeView(node, view, getPos);
        },
        figureCitation: (node, view, getPos) => {
          return new FigureCitationNodeView(node, view, getPos);
        }
      },
      handleDOMEvents: {
        focus: ({ state }: EditorView) => {
          if (this.props.onFocus && !this.props.isActive) {
            this.props.onFocus(state, this.props.name);
          }
          return true;
        },
        blur: ({ state }: EditorView) => {
          if (this.props.onBlur && this.props.isActive) {
            this.props.onBlur(state, this.props.name);
          }
          return true;
        }
      }
    };
  }

  focus() {
    this.focusEditor();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps: RichTextEditorProps) {
    // editor state is kept in sync between app state and editor state. App state will change when using formatting
    // or any toolbar menu. In this case when change comes from outside it needs to override editor state
    if (!nextProps.editorState.doc.eq(this.props.editorState.doc)) {
      this.updateEditorState(nextProps.editorState);
    }

    return (
      this.props.label !== nextProps.label ||
      this.props.isActive !== nextProps.isActive ||
      this.props.variant !== nextProps.variant
    );
  }

  componentDidUpdate() {
    // when component updates we need to restore focus
    this.restoreFocus();
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  updateEditorState(editorState: EditorState) {
    if (editorState !== this.editorView.state) {
      this.editorView.updateState(editorState);
    }
  }

  render() {
    return (
      <SectionContainer
        label={this.props.label}
        className={get(this.props, 'classes.root', '')}
        focused={this.props.isActive}
        id={this.props.id}
        variant={this.props.variant}
      >
        <div ref={this.createEditorView} className="prosemirrorContainer" />
      </SectionContainer>
    );
  }

  private async handleChange(change: Transaction) {
    const newState = this.editorView.state.apply(change);
    this.updateEditorState(newState);
    this.props.onChange(change, this.props.name);
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const clipboardSerializer = DOMSerializer.fromSchema(this.props.editorState.schema);
      Object.entries(this.props.editorState.schema.nodes).forEach(([nodeName, nodeType]) => {
        clipboardSerializer.nodes[nodeName] = get(nodeType, 'spec.toClipboardDOM', clipboardSerializer.nodes[nodeName]);
      });

      const additionalOptions = this.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        clipboardSerializer,
        handleScrollToSelection: (view: EditorView) => {
          return this.isFocusControlDelegated();
        },
        clipboardTextSerializer,
        state: this.props.editorState,
        dispatchTransaction: this.handleChange
      });
    }
  };

  private focusEditor() {
    const { from, to } = this.editorView.state.selection;
    restoreSelection(this.editorView, from, to);
    this.editorView.focus();
  }

  private isFocusControlDelegated() {
    const { $from } = this.editorView.state.selection;
    return this.editorView.dom.contains(document.activeElement) || hasParentNodeOf($from, ['boxText', 'figure']);
  }

  // focus is restored based on 3 conditions
  // #1 - component is active acc. to application state (prop: isActive)
  // #2 - editor does not have focus
  // #3 - focus is not delegated to a node view
  private restoreFocus() {
    if (this.props.isActive && this.editorView && !this.editorView.hasFocus()) {
      if (!this.isFocusControlDelegated()) {
        this.focusEditor();
      }
    }
  }
}

function clipboardTextSerializer(slice: Slice): string {
  const content = slice.content;
  const text = [];
  let separated = true;
  const blockSeparator = ' ';
  content.nodesBetween(0, content.size, (node, pos) => {
    if (node.isText) {
      text.push(node.text);
      separated = !blockSeparator;
    } else if (node.isLeaf && node.type.spec.toClipboardText) {
      text.push(node.type.spec.toClipboardText(node));
      separated = !blockSeparator;
    } else if (!separated && node.isBlock) {
      text.push(blockSeparator);
      separated = true;
    }
    return true;
  });

  return text.join('');
}
