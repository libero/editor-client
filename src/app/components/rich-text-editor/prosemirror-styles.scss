@import '../../styles/constants.module';

.prosemirrorContainer {
  line-height: initial;
  color: $text-primary;

  .ProseMirror .placeholder {
    color: rgba(0,0,0,0.38);
    font-weight: normal;
    pointer-events: none;
    height: 0;
  }

  .ProseMirror:focus .placeholder {
    display: none;
  }

  .selection-marker {
    background: #338FFF;
    color: white;
  }

  .ProseMirror {
    ::selection {
      background: transparent;
    }

    outline: none;

    h1 {
      font-size: 26px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h2 {
      font-size: 24px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h3 {
      font-size: 22px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h4 {
      font-size: 20px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    h5 {
      font-size: 18px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    h6 {
      font-size: 16px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    p {
      font-size: 16px;
      line-height: 1.5;
      margin: 0 0 16px 0;
      &:last-child {
        margin-bottom: 0;
      }
    }

    h1.article-title {
      margin: 0;
      padding: 0;
    }
  }
}
