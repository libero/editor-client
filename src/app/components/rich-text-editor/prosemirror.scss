.ProseMirror {
  .article-title {
    font-size: 36px;
    font-weight: 500;
  }
}