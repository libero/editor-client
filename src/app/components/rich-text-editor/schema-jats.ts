/* eslint-disable */
import { DOMOutputSpec, Schema, SchemaSpec } from 'prosemirror-model';

const pDOM = ['p', 0] as DOMOutputSpec,
  hrDOM = ['hr'] as DOMOutputSpec,
  preDOM = ['pre', ['code', 0]] as DOMOutputSpec,
  brDOM = ['br'] as DOMOutputSpec;

// :: Object
// [Specs](#model.NodeSpec) for the nodes defined in this schema.
export const nodes = {
  // :: NodeSpec The top level document node.
  doc: {
    content: 'block+'
  },

  article: {
    content: 'body*',
    parseDOM: [{ tag: 'article' }],
    toDOM() {
      return ['div'] as DOMOutputSpec;
    }
  },

  body: {
    content: 'block*',
    parseDOM: [{ tag: 'body' }],
    toDOM() {
      return ['div', 0] as DOMOutputSpec;
    }
  },

  boxed_text: {
    content: 'paragraph*',
    group: 'block',
    parseDOM: [{ tag: 'boxed-text' }],
    toDOM() {
      return ['div', { class: 'sc-card' }, 0] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A plain paragraph textblock. Represented in the DOM
  // as a `<p>` element.
  paragraph: {
    content: 'inline*',
    group: 'block',
    parseDOM: [{ tag: 'p' }],
    toDOM() {
      return pDOM;
    }
  },

  // :: NodeSpec A blockquote (`<blockquote>`) wrapping one or more blocks.
  blockquote: {
    content: 'block+',
    group: 'block',
    defining: true,
    parseDOM: [{ tag: 'blockquote' }],
    toDOM() {
      return ['div'] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A horizontal rule (`<hr>`).
  horizontal_rule: {
    group: 'block',
    parseDOM: [{ tag: 'hr' }],
    toDOM() {
      return hrDOM;
    }
  },

  // :: NodeSpec A heading textblock, with a `level` attribute that
  // should hold the number 1 to 6. Parsed and serialized as `<h1>` to
  // `<h6>` elements.
  heading: {
    attrs: { level: { default: 1 } },
    content: 'inline*',
    group: 'block',
    defining: true,
    parseDOM: [
      { tag: 'h1', attrs: { level: 1 } },
      { tag: 'h2', attrs: { level: 2 } },
      { tag: 'h3', attrs: { level: 3 } },
      { tag: 'h4', attrs: { level: 4 } },
      { tag: 'h5', attrs: { level: 5 } },
      { tag: 'h6', attrs: { level: 6 } }
    ],
    toDOM(node) {
      return ['h' + node.attrs.level, 0] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A code listing. Disallows marks or non-text inline
  // nodes by default. Represented as a `<pre>` element with a
  // `<code>` element inside of it.
  code_block: {
    content: 'text*',
    marks: '',
    group: 'block',
    code: true,
    defining: true,
    parseDOM: [{ tag: 'pre', preserveWhitespace: 'full' }],
    toDOM() {
      return preDOM;
    }
  },

  // :: NodeSpec The text node.
  text: {
    group: 'inline'
  },

  // :: NodeSpec An inline image (`<img>`) node. Supports `src`,
  // `alt`, and `href` attributes. The latter two default to the empty
  // string.
  image: {
    inline: true,
    attrs: {
      src: {},
      alt: { default: null },
      title: { default: null }
    },
    group: 'inline',
    draggable: true,
    parseDOM: [
      {
        tag: 'img[src]',
        getAttrs(dom) {
          return {
            src: dom.getAttribute('src'),
            title: dom.getAttribute('title'),
            alt: dom.getAttribute('alt')
          };
        }
      }
    ],
    toDOM(node) {
      const { src, alt, title } = node.attrs;
      return ['img', { src, alt, title }] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A hard line break, represented in the DOM as `<br>`.
  hard_break: {
    inline: true,
    group: 'inline',
    selectable: false,
    parseDOM: [{ tag: 'br' }],
    toDOM() {
      return brDOM;
    }
  },

  ext_link: {
    inline: true,
    content: 'inline*',
    attrs: {
      type: { default: 'uri' },
      href: {}
    },
    group: 'inline',
    parseDOM: [
      {
        tag: 'ext-link',
        getAttrs(dom) {
          return {
            type: dom.getAttribute('ext-link-type'),
            href: dom.getAttribute('xlink:href')
          };
        }
      }
    ],
    toDOM(node) {
      const { href } = node.attrs;
      return ['a', { href }, 0] as DOMOutputSpec;
    }
  },

  //<xref ref-type="bibr" rid="bib5">Katz et al., 2009</xref>
  xref: {
    inline: true,
    content: 'inline*',
    attrs: {
      type: { default: 'bibr' },
      href: { default: '#' }
    },
    group: 'inline',
    parseDOM: [
      {
        tag: 'xref',
        getAttrs(dom) {
          return {
            type: dom.getAttribute('ref-type'),
            href: dom.getAttribute('rid')
          };
        }
      }
    ],
    toDOM(node) {
      const { href } = node.attrs;
      return ['a', { href }, 0] as DOMOutputSpec;
    }
  },

  figure: {
    content: 'label* caption* graphic* attrib*',
    group: 'block',
    attrs: {
      id: {},
      position: { default: 'float' }
    },
    parseDOM: [
      {
        tag: 'fig',
        getAttrs(dom) {
          return {
            id: dom.getAttribute('id'),
            position: dom.getAttribute('position')
          };
        }
      }
    ],
    toDOM(node) {
      return ['div', { class: 'sc-card' }, 0] as DOMOutputSpec;
    }
  },

  label: {
    content: 'inline*',
    group: 'figure',
    selectable: false,
    parseDOM: [
      {
        tag: 'label'
      }
    ],
    toDOM(node) {
      return ['h2', 0] as DOMOutputSpec;
    }
  },

  caption: {
    content: 'title* paragraph*',
    group: 'figure',
    parseDOM: [
      {
        tag: 'caption'
      }
    ],
    toDOM() {
      return ['div', 0] as DOMOutputSpec;
    }
  },

  title: {
    content: 'text*',
    parseDOM: [
      {
        tag: 'title'
      }
    ],
    toDOM() {
      return ['h3', 0] as DOMOutputSpec;
    }
  },

  graphic: {
    attrs: {
      src: {}
    },
    group: 'figure',
    draggable: true,
    parseDOM: [
      {
        tag: 'graphic',
        getAttrs(dom) {
          return {
            src: dom.getAttribute('xlink:href')
          };
        }
      }
    ],
    toDOM(node) {
      const { src } = node.attrs;
      return ['img', { src }] as DOMOutputSpec;
    }
  },

  attrib: {
    content: 'text*',
    parseDOM: [
      {
        tag: 'attrib'
      }
    ],
    toDOM(node) {
      return ['h3', 0] as DOMOutputSpec;
    }
  }
};

const emDOM = ['em', 0] as DOMOutputSpec,
  strongDOM = ['strong', 0] as DOMOutputSpec,
  codeDOM = ['code', 0] as DOMOutputSpec;

// :: Object [Specs](#model.MarkSpec) for the marks in the schema.
export const marks = {
  // :: MarkSpec A link. Has `href` and `title` attributes. `title`
  // defaults to the empty string. Rendered and parsed as an `<a>`
  // element.
  link: {
    attrs: {
      href: {},
      title: { default: null }
    },
    inclusive: false,
    parseDOM: [
      {
        tag: 'a[href]',
        getAttrs(dom) {
          return { href: dom.getAttribute('href'), title: dom.getAttribute('title') };
        }
      }
    ],
    toDOM(node) {
      const { href, title } = node.attrs;
      return ['a', { href, title }, 0] as DOMOutputSpec;
    }
  },

  // :: MarkSpec An emphasis mark. Rendered as an `<em>` element.
  // Has parse rules that also match `<i>` and `font-style: italic`.
  em: {
    parseDOM: [{ tag: 'i' }, { tag: 'italic' }, { tag: 'em' }, { style: 'font-style=italic' }],
    toDOM() {
      return emDOM;
    }
  },

  // :: MarkSpec A strong mark. Rendered as `<strong>`, parse rules
  // also match `<b>` and `font-weight: bold`.
  strong: {
    parseDOM: [
      { tag: 'strong' },
      { tag: 'bold' },
      // This works around a Google Docs misbehavior where
      // pasted content will be inexplicably wrapped in `<b>`
      // tags with a font-weight normal.
      { tag: 'b', getAttrs: (node) => node.style.fontWeight !== 'normal' && null },
      { style: 'font-weight', getAttrs: (value) => /^(bold(er)?|[5-9]\d{2,})$/.test(value) && null }
    ],
    toDOM() {
      return strongDOM;
    }
  },

  // :: MarkSpec Code font mark. Represented as a `<code>` element.
  code: {
    parseDOM: [{ tag: 'code' }],
    toDOM() {
      return codeDOM;
    }
  }
};

// :: Schema
// This schema roughly corresponds to the document schema used by
// [CommonMark](http://commonmark.org/), minus the list elements,
// which are defined in the [`prosemirror-schema-list`](#schema-list)
// module.
//
// To reuse elements from this schema, extend or read from its
// `spec.nodes` and `spec.marks` [properties](#model.Schema.spec).

type NodesType = keyof typeof nodes;
type MarksType = keyof typeof marks;
export const schema = new Schema<NodesType, MarksType>({ nodes, marks } as SchemaSpec<NodesType, MarksType>);
/* eslint-enable */
