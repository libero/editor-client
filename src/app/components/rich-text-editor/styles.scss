.editorview-wrapper {
  width: 100%;
  height: 100%;
  position: relative;
  padding: 8px;
  border: solid 1px rgba(0, 0, 0, 0.23);
  border-radius: 4px;

  .editor-label {
    font-size: 12px;
    color: #263238;
    padding: 0 4px;
  }
}
