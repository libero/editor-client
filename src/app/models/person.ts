import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';
import { history } from 'prosemirror-history';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';

import * as bioConfig from 'app/models/config/author-bio.config';
import { Affiliation } from 'app/models/affiliation';
import { getTextContentFromPath, makeSchemaFromConfig } from 'app/models/utils';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Person extends BackmatterEntity {
  firstName: string;
  lastName: string;
  suffix?: string;
  isAuthenticated?: boolean;
  orcid?: string;
  email?: string;
  bio?: EditorState;
  isCorrespondingAuthor?: boolean;
  affiliations?: string[];
  hasCompetingInterest?: boolean;
  competingInterestStatement?: string;

  constructor(data?: JSONObject | Element, notesXml?: Element) {
    super();
    this.createEntity(data);
    if (data instanceof Element && notesXml instanceof Element) {
      this.setCompetingInterests(data, notesXml);
    }
  }

  public getDisplayName(): string {
    return [this.firstName, this.lastName, this.suffix].filter((_) => _).join(' ');
  }

  public getAffiliationsLabels(affiliations: Affiliation[]): string[] {
    return this.affiliations
      .map((affiliationId) => {
        const affiliation = affiliations.find(({ id }) => id === affiliationId);
        return affiliation ? affiliation.label : undefined;
      })
      .filter(Boolean)
      .sort((a, b) => Number(a) - Number(b));
  }

  clone(): Person {
    const person = new Person();
    person.firstName = this.firstName;
    person.lastName = this.lastName;
    person.suffix = this.suffix;
    person.isAuthenticated = this.isAuthenticated;
    person.email = this.email;
    person.bio = this.bio;
    person.isCorrespondingAuthor = this.isCorrespondingAuthor;
    person.affiliations = [...this.affiliations];
    person.hasCompetingInterest = this.hasCompetingInterest;
    person.competingInterestStatement = this.competingInterestStatement;

    return person;
  }

  protected fromXML(xml: Element): void {
    const orcidEl = xml.querySelector('contrib-id[contrib-id-type="orcid"]');

    this._id = xml.getAttribute('id') || this._id;
    this.firstName = getTextContentFromPath(xml, 'name > given-names');
    this.lastName = getTextContentFromPath(xml, 'name > surname');
    this.suffix = getTextContentFromPath(xml, 'name > suffix');
    this.isAuthenticated = orcidEl ? orcidEl.getAttribute('authenticated') === 'true' : false;
    this.orcid = orcidEl ? this.getOrcid(orcidEl.textContent) : '';
    this.bio = this.createBioEditorStateFromXml(xml.querySelector('bio'));
    this.email = getTextContentFromPath(xml, 'email');
    this.isCorrespondingAuthor = xml.getAttribute('corresp') === 'yes';
    this.affiliations = Array.from(xml.querySelectorAll('xref[ref-type="aff"]')).map((xRef) =>
      xRef.getAttribute('rid')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this._id = (json.id as string) || this._id;
    this.firstName = json.firstName as string;
    this.lastName = json.lastName as string;
    this.suffix = json.suffix as string;
    this.isAuthenticated = json.isAuthenticated as boolean;
    this.orcid = json.orcid as string;
    this.bio = json.bio
      ? this.createBioEditorStateFromJSON(json.bio as JSONObject)
      : this.createBioEditorStateFromXml();
    this.email = json.email as string;
    this.isCorrespondingAuthor = json.isCorrespondingAuthor as boolean;
    this.affiliations = (json.affiliations as string[]) || [];
  }

  protected createBlank(): void {
    this.firstName = '';
    this.lastName = '';
    this.suffix = '';
    this.isAuthenticated = false;
    this.orcid = '';
    this.bio = this.createBioEditorStateFromXml();
    this.email = '';
    this.isCorrespondingAuthor = false;
    this.affiliations = [];
  }

  private getOrcid(orcidUrl: string): string {
    const matches = orcidUrl.match(/(([0-9]{4}-?){4})/g);
    if (matches) {
      return matches[0];
    }
    return '';
  }

  private setCompetingInterests(dataXml: Element, notesXml: Element): void {
    const competingInterestEl = Array.from(notesXml.querySelectorAll('[fn-type="COI-statement"]')).find(
      (fnEl: Element) => {
        const id = fnEl.getAttribute('id');
        return dataXml.querySelector(`xref[ref-type="fn"][rid="${id}"]`);
      }
    );

    this.hasCompetingInterest = competingInterestEl
      ? competingInterestEl.textContent !== 'No competing interests declared'
      : false;
    this.competingInterestStatement = competingInterestEl ? competingInterestEl.textContent : '';
  }

  private createBioEditorStateFromXml(bio?: Element): EditorState {
    const schema = makeSchemaFromConfig(bioConfig.topNode, bioConfig.nodes, bioConfig.marks);
    return EditorState.create({
      doc: bio ? ProseMirrorDOMParser.fromSchema(schema).parse(bio) : undefined,
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), history()]
    });
  }

  private createBioEditorStateFromJSON(json: JSONObject): EditorState {
    const blankState = this.createBioEditorStateFromXml();
    return EditorState.fromJSON(
      {
        schema: blankState.schema,
        plugins: blankState.plugins
      },
      json
    );
  }
}

export function createAuthorsState(authorsXml: Element[], notesXml: Element | undefined): Person[] {
  return authorsXml.map((xml) => new Person(xml, notesXml));
}
