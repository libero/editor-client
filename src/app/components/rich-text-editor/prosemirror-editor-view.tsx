import React from 'react';
import { EditorState, Transaction } from 'prosemirror-state';
import { EditorProps, EditorView } from 'prosemirror-view';
import classnames from 'classnames';
import 'prosemirror-view/style/prosemirror.css';

import './prosemirror-styles.scss';

interface ProseMirrorEditorViewState {
  editorState: EditorState;
}

export interface ProseMirrorEditorViewProps {
  className?: string;
  editorState: EditorState;
  options?: Partial<EditorProps>;
  onChange: (tx: Transaction) => void;
}

export class ProseMirrorEditorView extends React.Component<ProseMirrorEditorViewProps, ProseMirrorEditorViewState> {
  public props;
  public editorView: EditorView;

  focus() {
    // debugger;
    this.editorView.focus();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps.editorState !== this.editorView.state) {
      this.editorView.updateState(nextProps.editorState);
    }
    return false;
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  render() {
    // Render just an empty div which is then used as a container for an
    // EditorView instance.
    return <div ref={this.createEditorView} className={classnames('prosemirrorContainer', this.props.className)} />;
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const additionalOptions = this.props.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        state: this.props.editorState,
        dispatchTransaction: (tx: Transaction) => this.props.onChange(tx)
      });
    }
  };
}
