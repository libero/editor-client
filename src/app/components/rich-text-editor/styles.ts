import { makeStyles } from '@material-ui/core/styles';

export const useRichTextEditorStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    height: '100%',
    padding: theme.spacing(1, 1.5, 1.5, 1.5),
    border: `solid 1px`,
    borderColor: theme.palette.secondary.dark,
    borderRadius: theme.spacing(0.5),
    boxSizing: 'border-box'
  },

  label: {
    fontSize: '0.75rem',
    color: theme.palette.text.secondary,
    padding: theme.spacing(0, 0.5)
  },

  focused: {
    borderColor: theme.palette.primary.light,
    '& > legend': {
      color: theme.palette.primary.light
    }
  }
}));
