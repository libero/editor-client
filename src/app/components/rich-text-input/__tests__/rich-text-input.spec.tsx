import React from 'react';
import { EditorState } from 'prosemirror-state';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';

import { Person } from '../../../models/person';
import { RichTextInput } from '../';
import { ProseMirrorEditorView } from '../../rich-text-editor/prosemirror-editor-view';

jest.mock('@material-ui/lab', () => ({
  ToggleButtonGroup: ({ children }) => <div data-cmp="ToggleButtonGroup">{children}</div>,
  ToggleButton: ({ children }) => <div data-cmp="ToggleButton">{children}</div>
}));

jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn(),
    makeStyles: jest.requireActual('@material-ui/core/styles').makeStyles
  };
});

describe('RichTextInput', () => {
  it('should render component', () => {
    const wrapper = create(
      <RichTextInput
        editorState={givenEditorState()}
        name="test-editor-input"
        label="Test Editor Input"
        onChange={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle an event', () => {
    const onChangeHandler = jest.fn();
    const editorState = givenEditorState();
    const wrapper = mount(
      <RichTextInput
        editorState={editorState}
        name="test-editor-input"
        label="Test Editor Input"
        onChange={onChangeHandler}
      />
    );
    const change = editorState.tr;
    const prosemirrorComponent = wrapper.find(ProseMirrorEditorView).at(0);
    (prosemirrorComponent.instance() as ProseMirrorEditorView).editorView.dispatch(change);

    expect(onChangeHandler).toHaveBeenCalledWith(change);
  });

  function givenEditorState(): EditorState {
    const authorsContainer = document.createElement('div');
    authorsContainer.innerHTML = `
        <name><surname>Atherden</surname><given-names>Fred</given-names><suffix>Capt.</suffix></name>
        <contrib-id authenticated="true" contrib-id-type="orcid">https://orcid.org/0000-0002-6048-1470</contrib-id>
        <email>f.atherden@elifesciences.org</email>
        <bio>
            <p><bold>Fred Atherden</bold> is in the Production Department, eLife Sciences, Cambridge, United Kingdoms
                <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution License</ext-link>
            </p> 
        </bio>`;
    const person = new Person(authorsContainer);
    return person.bio;
  }
});
