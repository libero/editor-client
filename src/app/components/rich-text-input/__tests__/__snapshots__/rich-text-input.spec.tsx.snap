// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichTextInput should render component 1`] = `
<div
  className="makeStyles-container-1 makeStyles-root-4 makeStyles-outlinedVariant-5"
>
  <label
    className="makeStyles-label-7"
  >
    Test Editor Input
  </label>
  <div
    data-cmp="ToggleButtonGroup"
  />
  <div
    className="prosemirrorContainer"
  />
</div>
`;
